---
title: "R Notebook"
output: 
---

```{r message=FALSE}
options(scipen=999)
library(tidyverse)
library(data.table)
library(R.utils)
library(RSQLite)
library(BSgenome.Hsapiens.1000genomes.hs37d5)
```

# Load Splice Junction stats data

```{r}
intropolis_sj <- fread('output/intropolis_annotated.tsv.gz', nThread = 4)
gtex_sj <- fread('output/gtex_annotated.tsv.gz', nThread = 4)
setnames(gtex_sj, old = c('n_samples'), new = c('sample_count'))
```

# Check for duplicates
```{r}
if (nrow(intropolis_sj[,.(N = .N), by = .(chr, donor_pos, acceptor_pos)][N > 1]) != 0) { stop("Duplicate rows in intropolis data") }
if (nrow(gtex_sj[,.(N = .N), by = .(chr, donor_pos, acceptor_pos)][N > 1]) != 0) { stop("Duplicate rows in GTEx data") }
```

# Load Exons reference

```{r}
exons <- fread('data/ensembl_nasty_exons.tsv.gz')
exons[, donor_pos := ifelse(strand == '+', region_end + 1, region_start - 1)]
exons[, acceptor_pos := ifelse(strand == '+', region_start - 1, region_end + 1)]

exons[, nextExon_acceptorPos :=  lead(acceptor_pos), by = .(tx_id)]
exons[, prevExon_donorPos :=  lag(donor_pos), by = .(tx_id)]

exons[, strand_mult := ifelse(strand == '+', 1, -1)]

# remove single exon transcripts
exons <- exons[!is.na(nextExon_acceptorPos) | !is.na(prevExon_donorPos)]

# for first and last exons annotate neighbouring splice sites slighlty differently
exons[is.na(nextExon_acceptorPos), nextExon_acceptorPos := donor_pos]
exons[is.na(prevExon_donorPos), prevExon_donorPos := acceptor_pos]

# get bounds of transcript
exons[, tx_start := acceptor_pos[region_no == 1], by = tx_id]
exons[, tx_end := donor_pos[region_no == nexon], by = tx_id]


# make helper columns
exons[, ES_d_lb := min(donor_pos,tx_end), by = 1:nrow(exons)]
exons[, ES_d_ub := max(donor_pos,tx_end), by = 1:nrow(exons)]

exons[, ES_a_lb := min(acceptor_pos,tx_start), by = 1:nrow(exons)]
exons[, ES_a_ub := max(acceptor_pos,tx_start), by = 1:nrow(exons)]

exons[, css_lb_d := min(acceptor_pos, nextExon_acceptorPos), by = 1:nrow(exons)]
exons[, css_ub_d := max(acceptor_pos, nextExon_acceptorPos), by = 1:nrow(exons)]

exons[, css_lb_a := min(donor_pos, prevExon_donorPos), by = 1:nrow(exons)]
exons[, css_ub_a := max(donor_pos, prevExon_donorPos), by = 1:nrow(exons)]

# filter to unique intron-exon pairs
exons_dup <- exons[, .(chrom, region_start, region_end, strand, nextExon_acceptorPos, prevExon_donorPos)]

exons_unique <- exons[!duplicated(exons_dup),]
```


```{r}
get_missplicing_table_donors <- function(exons_df, sj_df, col_sel) {
  # step 1. get normal splicing
  NS <- sj_df[exons_df, 
              on = .(chr, donor_pos = exon_donor_pos, acceptor_pos = nextExon_acceptorPos, strand)]
  NS[is.na(splicing_event_class), splicing_event_class := 'normal splicing']
  NS[,exon_donor_pos := donor_pos ]
  # step 2. get exon skipping events
  ## anchored at previous donor 
  ES_d <- sj_df[annotated_acceptor == 1][exons_df,allow.cartesian=TRUE,on = .(chr, strand, 
                                                         donor_pos = prevExon_donorPos)
                                         ][data.table::between(acceptor_pos,ES_d_lb,ES_d_ub,incbounds = F)]
  

  ES_d <- exons[, .(tx_id, donor_pos, donor_exon_no = region_no)][ES_d, on = .(tx_id, donor_pos)]
  ES_d <- exons[, .(tx_id, acceptor_pos, acceptor_exon_no = region_no)][ES_d, on = .(tx_id, acceptor_pos)]
  
  ## anchored at next acceptor
  ES_a <- sj_df[annotated_donor == 1][exons_df,allow.cartesian=TRUE,on = .(chr, strand, 
                                                      acceptor_pos = nextExon_acceptorPos)
                                      ][data.table::between(donor_pos,ES_a_lb,
                                                ES_a_ub,incbounds = F)]
  
  ES_a <- exons[, .(tx_id, donor_pos, donor_exon_no = region_no)][ES_a, on = .(tx_id, donor_pos)]
  ES_a <- exons[, .(tx_id, acceptor_pos, acceptor_exon_no = region_no)][ES_a, on = .(tx_id, acceptor_pos)]

  # get exon skipping info - exons skipped, #nts omitted from transcript ( to calculate frame)
  ES <- rbind(ES_a, ES_d, fill = T)
  ES <- ES[!is.na(acceptor_exon_no) & !is.na(donor_exon_no)]
  ES[, skipped_exons_count := acceptor_exon_no - donor_exon_no - 1]
  
  ES[, paste_first := donor_exon_no + 1]
  ES[, paste_second := acceptor_exon_no - 1]
  
  nts_skipped_join <- exons[, .(tx_id, region_no, region_width)][ES[, .(tx_id, exon_donor_pos,exon_acceptor_pos, 
                                                                        paste_first, paste_second)], 
                                                                 on = .(tx_id, region_no >= paste_first, 
                                                                        region_no <= paste_second)]
  
  nts_skipped_join <- nts_skipped_join[, .(skipped_exons_nt = sum(region_width)), 
                                       by = .(tx_id, exon_donor_pos, exon_acceptor_pos, 
                                              paste_first = region_no, paste_second = region_no.1)]
  
  ES <- nts_skipped_join[ES, on = .(tx_id, exon_donor_pos,exon_acceptor_pos, 
                                    paste_first, paste_second)]
  
  ES[paste_second == paste_first, paste_second := NA]
  ES[, skipped_exons_id := do.call(paste, c(.SD, sep="-")), .SDcols= paste_first:paste_second]
  ES[, skipped_exons_id := gsub('-NA', '',skipped_exons_id )]
  
  ES[splicing_event_class == 'normal splicing', splicing_event_class := 'exon skipping (annotated)']
  ES[, missplicing_inframe := ifelse(skipped_exons_nt %% 3 == 0, 1, 0)]

  # step 3. get cryptic splicing events
  ## cryptic donor
  CSS <-sj_df[exons_df,allow.cartesian = TRUE, on = .(chr, strand, 
                                                     acceptor_pos = nextExon_acceptorPos)
              ][donor_pos != exon_donor_pos & 
                  data.table::between(donor_pos, css_lb_d, css_ub_d)]
 
  # calculate distance of cryptic to authentic splice site
  CSS[, cryptic_pos := donor_pos]
  CSS[, cryptic_distance := strand_mult * (donor_pos - exon_donor_pos) ]
  CSS[, missplicing_inframe := ifelse(cryptic_distance %% 3 == 0, 1, 0)]
  CSS[, cryptic_distance := ifelse(cryptic_distance > 0, cryptic_distance + 1, cryptic_distance)]
  CSS[splicing_event_class == 'normal splicing' | annotated_donor == 1, splicing_event_class := 'alternative donor (annotated)']

  
  # step 4. combine into one mis-splicing event table
  ms_table <- rbind(NS, ES, CSS, fill = T)
  setDT(ms_table)
  ms_table[, splice_site_pos := exon_donor_pos]
  cols <- c('gene_name', 'tx_id','canonical', 'chr', 'splice_site_pos',
            'exon_donor_pos', 'exon_acceptor_pos', 
            'region_width', 'region_no', 'strand',
            'donor_pos', 'acceptor_pos', 
            'sample_count', col_sel, 'splicing_event_class',
            'skipped_exons_count', 'skipped_exons_id', 'skipped_exons_nt',
            'cryptic_pos', 'cryptic_distance', 'missplicing_inframe')
  ms_table <- distinct(ms_table[, ..cols])
  
  setorder(ms_table, tx_id, region_no, -sample_count)
  
  
  return(ms_table)
}
```

```{r}
get_missplicing_table_acceptors <- function(exons_df, sj_df, col_sel) {
  # step 1. get normal splicing
  NS <- sj_df[exons_df, 
              on = .(chr, donor_pos = prevExon_donorPos, acceptor_pos = exon_acceptor_pos, strand)]
  NS[is.na(splicing_event_class), splicing_event_class := 'normal splicing']
  NS[, exon_acceptor_pos := acceptor_pos]
  
  # step 2. get exon skipping events
  ## anchored at previous donor 
  ES_d <- sj_df[annotated_acceptor == 1][exons_df,allow.cartesian=TRUE,on = .(chr, strand, 
                                                         donor_pos = prevExon_donorPos)
                                         ][data.table::between(acceptor_pos,ES_d_lb,ES_d_ub,incbounds = F)]
  
  ES_d <- exons[, .(tx_id, donor_pos, donor_exon_no = region_no)][ES_d, on = .(tx_id, donor_pos)]
  ES_d <- exons[, .(tx_id, acceptor_pos, acceptor_exon_no = region_no)][ES_d, on = .(tx_id, acceptor_pos)]
  
  ## anchored at next acceptor
  ES_a <- sj_df[annotated_donor == 1][exons_df,allow.cartesian=TRUE,on = .(chr, strand, 
                                                      acceptor_pos = nextExon_acceptorPos)
                                      ][data.table::between(donor_pos,ES_a_lb,
                                                ES_a_ub,incbounds = F)]
  
  ES_a <- exons[, .(tx_id, donor_pos, donor_exon_no = region_no)][ES_a, on = .(tx_id, donor_pos)]
  ES_a <- exons[, .(tx_id, acceptor_pos, acceptor_exon_no = region_no)][ES_a, on = .(tx_id, acceptor_pos)]
  
  # get exon skipping info - exons skipped, #nts omitted from transcript ( to calculate frame)
  ES <- rbind(ES_a, ES_d, fill = T)
  ES <- ES[!is.na(acceptor_exon_no) & !is.na(donor_exon_no)]
  ES[, skipped_exons_count := acceptor_exon_no - donor_exon_no - 1]
  
  ES[, paste_first := donor_exon_no + 1]
  ES[, paste_second := acceptor_exon_no - 1]
  
  nts_skipped_join <- exons[, .(tx_id, region_no, region_width)][ES[, .(tx_id, exon_donor_pos,exon_acceptor_pos, 
                                                                        paste_first, paste_second)], 
                                                                 on = .(tx_id, region_no >= paste_first, region_no <= paste_second)]
  
  nts_skipped_join <- nts_skipped_join[, .(skipped_exons_nt = sum(region_width)), 
                                       by = .(tx_id, exon_donor_pos, exon_acceptor_pos, paste_first = region_no, paste_second = region_no.1)]
  
  ES <- nts_skipped_join[ES, on = .(tx_id, exon_donor_pos,exon_acceptor_pos, 
                                    paste_first, paste_second)]
  
  ES[paste_second == paste_first, paste_second := NA]
  ES[, skipped_exons_id := do.call(paste, c(.SD, sep="-")), .SDcols= paste_first:paste_second]
  ES[, skipped_exons_id := gsub('-NA', '',skipped_exons_id )]
  
  ES[splicing_event_class == 'normal splicing', splicing_event_class := 'exon skipping (annotated)']
  ES[, missplicing_inframe := ifelse(skipped_exons_nt %% 3 == 0, 1, 0)]
  
  # step 3. get cryptic splicing events
  ## cryptic acceptor
  CSS <-sj_df[exons_df,allow.cartesian=TRUE, on = .(chr, strand, 
                                                     donor_pos = prevExon_donorPos)
              ][acceptor_pos != exon_acceptor_pos & 
                  data.table::between(acceptor_pos, css_lb_a, css_ub_a)]
  
  # calculate distance of cryptic to authentic splice site
  CSS[, cryptic_pos := acceptor_pos]
  CSS[, cryptic_distance := strand_mult * (acceptor_pos - exon_acceptor_pos) ]
  CSS[, missplicing_inframe := ifelse(cryptic_distance %% 3 == 0, 1, 0)]
  CSS[, cryptic_distance := ifelse(cryptic_distance < 0, cryptic_distance - 1, cryptic_distance)]
  CSS[splicing_event_class == 'normal splicing'| annotated_acceptor == 1, splicing_event_class := 'alternative acceptor (annotated)']
  
  # step 4. combine into one mis-splicing event table
  ms_table <- rbind(NS, ES, CSS, fill = T)
  setDT(ms_table)
  ms_table[, splice_site_pos := exon_acceptor_pos]
  cols <- c('gene_name', 'tx_id','canonical', 'chr', 'splice_site_pos',
            'exon_donor_pos', 'exon_acceptor_pos', 
            'region_width', 'region_no', 'strand',
            'donor_pos', 'acceptor_pos', 
            'sample_count', col_sel, 'splicing_event_class',
            'skipped_exons_count', 'skipped_exons_id', 'skipped_exons_nt',
            'cryptic_pos', 'cryptic_distance', 'missplicing_inframe')
  ms_table <- distinct(ms_table[, ..cols])
  
 
  
  setorder(ms_table, tx_id, region_no, -sample_count)
  
  return(ms_table)
}
```


```{r}

gtex_columns_of_interest <- c('q99_uniq_map_reads','max_uniq_map_reads','q99_multi_map_reads','max_multi_map_reads')

exons_df_donor <- exons[last_region== 0, 
                               .(gene_name, tx_id,canonical,
                                 chr = chrom, region_start, region_end, region_width, region_no, 
                                 nexon, strand, last_region,
                                 exon_donor_pos = donor_pos, 
                                 exon_acceptor_pos = acceptor_pos, 
                                 nextExon_acceptorPos, 
                                 prevExon_donorPos, 
                                 strand_mult, tx_start, tx_end, 
                                 ES_d_lb, ES_d_ub, ES_a_lb, ES_a_ub, 
                                 css_lb_d, css_ub_d, css_lb_a, css_ub_a)]

ms_intropolis_donors <- get_missplicing_table_donors(exons_df_donor,intropolis_sj, 'max_sj')
ms_intropolis_donors[, ss_type := 'donor']

ms_gtex_donors <- get_missplicing_table_donors(exons_df_donor,gtex_sj, gtex_columns_of_interest)
ms_gtex_donors[, ss_type := 'donor']

exons_df_acceptor <- exons[region_no != 1, 
                               .(gene_name, tx_id,canonical,
                                 chr = chrom, region_start, region_end, region_width, region_no, 
                                 nexon, strand, last_region,
                                 exon_donor_pos = donor_pos, 
                                 exon_acceptor_pos = acceptor_pos, 
                                 nextExon_acceptorPos, 
                                 prevExon_donorPos, 
                                 strand_mult, tx_start, tx_end, 
                                 ES_d_lb, ES_d_ub, ES_a_lb, ES_a_ub, 
                                 css_lb_d, css_ub_d, css_lb_a, css_ub_a)]

ms_intropolis_acceptors <- get_missplicing_table_acceptors(exons_df_acceptor,intropolis_sj, 'max_sj')
ms_intropolis_acceptors[, ss_type := 'acceptor']

ms_gtex_acceptors <- get_missplicing_table_acceptors(exons_df_acceptor,gtex_sj, gtex_columns_of_interest)
ms_gtex_acceptors[, ss_type := 'acceptor']
```

```{r}
ms_intropolis <- distinct(rbind(ms_intropolis_donors, ms_intropolis_acceptors))
ms_gtex <- distinct(rbind(ms_gtex_donors, ms_gtex_acceptors))
```


```{r}
# fwrite(ms_intropolis, 'output/missplicingPredictions_intropolis.tsv.gz', sep = '\t', nThread = 4, compress = 'gzip')
# fwrite(ms_gtex, 'output/missplicingPredictions_gtex.tsv.gz', sep = '\t', nThread = 4, compress = 'gzip')
```

# making combined table
```{r}
# ms_intropolis <- fread('output/missplicingPredictions_intropolis.tsv.gz', sep = '\t', nThread = 4)
# ms_gtex <- fread('output/missplicingPredictions_gtex.tsv.gz', sep = '\t', nThread = 4)
```

```{r}
# merging gtex and intropolis
setnames(ms_gtex, 'sample_count', 'gtex_sample_count')
setnames(ms_intropolis, 'sample_count', 'intropolis_sample_count')

ms_merge <- merge(ms_intropolis, ms_gtex, 
                  by = intersect(names(ms_intropolis), names(ms_gtex)), all = T)

# Clear up some mem
rm(ms_gtex_acceptors)
rm(ms_gtex_donors)
rm(ms_intropolis_acceptors)
rm(ms_intropolis_donors)

# get summary of sampel count & max sj across gtex and inntropolis
ms_merge[is.na(intropolis_sample_count), intropolis_sample_count := 0]
ms_merge[is.na(gtex_sample_count), gtex_sample_count := 0]
ms_merge[is.na(max_sj), max_sj := 0]
ms_merge[is.na(q99_uniq_map_reads), max_uniq_map_reads := 0]
ms_merge[is.na(q99_multi_map_reads), max_multi_map_reads := 0]

ms_merge[, sample_count := Reduce(`+`, .SD), .SDcol = c(21,23)]
ms_merge[, multimap_flag := ifelse(max_multi_map_reads > 0, 1, 0)]

ms_merge[, in_gtex := ifelse(gtex_sample_count > 0, TRUE, FALSE)]
ms_merge[, in_intropolis := ifelse(intropolis_sample_count > 0, TRUE, FALSE)]
 
setorder(ms_merge, splice_site_pos, tx_id, -sample_count)

ms_merge[, event_rank := rowid(splice_site_pos), by = list(tx_id)]

ms_merge[missplicing_inframe == 1, missplicing_inframe := TRUE]
ms_merge[missplicing_inframe == 0, missplicing_inframe := FALSE]
ms_merge[splicing_event_class == 'normal splicing', missplicing_inframe := TRUE]
ms_merge[, missplicing_inframe := as.logical(missplicing_inframe)]

ms_merge2 <- ms_merge[,.(splice_site_pos, 
                         gene_name,
                         tx_id, 
                         canonical,
                         ss_type,
                         exon_no=region_no,
                         strand, 
                         splicing_event_class, 
                         event_rank, 
                         in_gtex, 
                         in_intropolis, 
                         missplicing_inframe,
                         gtex_sample_count,
                         intropolis_sample_count,
                         sample_count,
                         gtex_q99_uniq_map_reads = q99_uniq_map_reads,
                         gtex_max_uniq_map_reads = max_uniq_map_reads,
                         intropolis_max_reads = max_sj, 
                         gtex_multimap_flag = multimap_flag,
                         skipped_exons_count,
                         skipped_exons_id,
                         cryptic_distance, 
                         chr, 
                         donor_pos, 
                         acceptor_pos, 
                         assembly = 'hg19',
                         transcript_type = 'ensembl')]

```

# Check for duplicates
```{r}
merged_dup_count <- nrow(ms_merge2[,.(N = .N), by = .(splice_site_pos, tx_id, chr, donor_pos, acceptor_pos)][N > 1])
if (merged_dup_count != 0) { stop(paste("Duplicate rows in merged data, n = ", merged_dup_count)) }
```
```{r}
ms_merge2[,.(N = .N), by = .(splice_site_pos, tx_id, chr, donor_pos, acceptor_pos)][N > 1]

ms_merge2[ms_merge2[,.(N = .N), by = .(splice_site_pos, tx_id, chr, donor_pos, acceptor_pos)][N > 1], on = .(splice_site_pos, tx_id, chr, donor_pos, acceptor_pos)]
```

```{r}
fwrite(ms_merge2, 'output/40KRNA.tsv.gz', sep = '\t', compress = 'gzip', nThread = 4)
```

# writing sqlite dtabase

```{r}
ms_gtex <- ms_gtex[, .(tx_id, splice_site_pos, 
                 Event = splicing_event_class,
                 Samples = gtex_sample_count,
                 MaxMReads = max_multi_map_reads,
                 Q99MReads = q99_multi_map_reads,
                 MaxUReads = max_uniq_map_reads,
                 Q99UReads = q99_uniq_map_reads,
                 SkippedExons = skipped_exons_id,
                 CrypDist = as.character(cryptic_distance),
                 InFrame = missplicing_inframe,
                 Chr = chr,
                 Donor = donor_pos,
                 Acceptor = acceptor_pos)]

ms_gtex[is.na(CrypDist), CrypDist := '']

ms_intropolis <- ms_intropolis[, .(tx_id, splice_site_pos, 
                 Event = splicing_event_class,
                 Samples = intropolis_sample_count, 
                 MaxReads = max_sj,
                 SkippedExons = skipped_exons_id,
                 CrypDist = as.character(cryptic_distance),
                 InFrame = missplicing_inframe,
                 Chr = chr,
                 Donor = donor_pos,
                 Acceptor = acceptor_pos)]

ms_intropolis[is.na(CrypDist), CrypDist := '']

ms_merge <- ms_merge2[, .(tx_id, splice_site_pos, 
                 Event = splicing_event_class,
                 InIntropolis = in_intropolis,
                 InGtex = in_gtex,
                 GtexSamples = gtex_sample_count, 
                 IntropSamples = intropolis_sample_count, 
                 GtexUReads = gtex_max_uniq_map_reads, 
                 GtexQ99UReads = gtex_q99_uniq_map_reads,
                 IntropMaxReads = intropolis_max_reads,
                 GtexMMapFlag = gtex_multimap_flag,
                 SkippedExons = skipped_exons_id,
                 CrypDist = as.character(cryptic_distance),
                 InFrame = missplicing_inframe,
                 Chr = chr,
                 Donor = donor_pos,
                 Acceptor = acceptor_pos)]

ms_merge[is.na(CrypDist), CrypDist := '']
```

```{r}
intropolis_sj_db_file <- "output/db/missplicingPredictions_intropolis.sqlite"
intropolis_sj_db <- dbConnect(RSQLite::SQLite(), intropolis_sj_db_file)

dbWriteTable(intropolis_sj_db, "sj", ms_intropolis, overwrite = TRUE, append = FALSE)
dbSendQuery(intropolis_sj_db, "CREATE INDEX idx_tx ON sj (tx_id);")
dbDisconnect(intropolis_sj_db)
```

```{r}
gtex_sj_db_file <- "output/db/missplicingPredictions_gtex.sqlite"
gtex_sj_db <- dbConnect(RSQLite::SQLite(), gtex_sj_db_file)

dbWriteTable(gtex_sj_db, "sj", ms_gtex, overwrite = TRUE, append = FALSE)
dbSendQuery(gtex_sj_db, "CREATE INDEX idx_tx ON sj (tx_id);")
dbDisconnect(gtex_sj_db)
```

```{r}
merge_sj_db_file <- "output/db/missplicingPredictions_merge.sqlite"
merge_sj_db <- dbConnect(RSQLite::SQLite(), merge_sj_db_file)

dbWriteTable(merge_sj_db, "sj", ms_merge, overwrite = TRUE, append = FALSE)
dbSendQuery(merge_sj_db, "CREATE INDEX idx_tx ON sj (tx_id);")
dbDisconnect(merge_sj_db)
```



